core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java

```diff
@@ -695,11 +690,16 @@ public boolean append(Object kbase, long koff, int klen, Object vbase, long voff
       assert (vlen % 8 == 0);
       assert (longArray != null);
 
-      if (numKeys == MAX_CAPACITY
-        // The map could be reused from last spill (because of no enough memory to grow),
-        // then we don't try to grow again if hit the `growthThreshold`.
-        || !canGrowArray && numKeys > growthThreshold) {
-        return false;
+      if (numKeys >= growthThreshold) {
+        if (longArray.size() / 2 == MAX_CAPACITY) {
```

Comment:
Is `MAX_CAPACITY` still the maximum number of keys as per documentation of it? If we can have `longArray.size() / 2 == MAX_CAPACITY` at most for the capacity, the actually `numKeys` should be `MAX_CAPACITY / 2`, because we need two long array entries per key, right?